src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonSerializerOptions.Caching.cs

```diff
@@ -74,9 +74,11 @@ internal void ClearCaches()
         private void InitializeCachingContext()
         {
             _cachingContext = TrackedCachingContexts.GetOrCreate(this);
-            if (IsInitializedForReflectionSerializer)
+            if (_typeInfoResolver != null)
             {
-                _cachingContext.Options.IsInitializedForReflectionSerializer = true;
+                Debug.Assert(_cachingContext.Options._typeInfoResolver != null || _typeInfoResolver == s_defaultTypeInfoResolver,
+                    "EqualityComparer should guarantee that null is equivalent to s_defaultTypeInfoResolver");
+                _cachingContext.Options._typeInfoResolver ??= s_defaultTypeInfoResolver;
```

Comment:
Why is this hardcoding the default resolver in the caching context? It might not match what is stored in 
 `_typeInfoResolver

Reply:
there are only two cases here:
- (both `_cachingContext.Options` and) `_typeInfoResolver` is `null` or `s_defaultTypeInfoResolver` effectively meaning the same thing
- `_typeInfoResolver` is not null and equal between the two

in the latter case the `s_defaultTypeInfoResolver` will not be assigned and in the first case `null` and `s_defaultTypeInfoResolver` mean the same thing. See the assert before